### PR TITLE
Fix python bindings

### PR DESCRIPTION
**Pull request by sigmundklaa**
Fixes the Python bindings that were broken in CSP 2.0, discussed in #322.

Summary:
* Added a small example script as the current example scripts are quite big and I figured a minimal example could be good to have.
* An optional argument `version` was added to `pycsp_init` to allow the user to specify CSP version.
* The address that is passed to `pycsp_init` was removed because it is no longer used. This can be readded for backwards compatibility, but from my understanding the bindings have not been working at all so this should not break any functioning codebase. Please correct me if that is wrong, and I will readd the argument and just not use it.
* `pycsp_socket` caused a segmentation fault due to a pointer to a stack allocated socket being used. This was changed to instead use pythons memory allocator for the allocation of the socket.

### CODE REVIEW

**Comment by yashi** [comment]
Hi @sigmundklaa thanks for the pr. I'll take a look in a week.

**Review by yashi — COMMENTED** [review]
~Would you please write in the commit log why you need to "remove address from pycsp init"?~

Sorry.  you wrote it in the first comment.

> The address that is passed to pycsp_init was removed because it is no longer used. This can be readded for backwards compatibility, but from my understanding the bindings have not been working at all so this should not break any functioning codebase. Please correct me if that is wrong, and I will readd the argument and just not use it.

File: wscript [hunk 1]
```diff
@@ -73,7 +73,7 @@ def configure(ctx):
 
     # Setup CFLAGS
     if (len(ctx.stack_path) <= 1) and (len(ctx.env.CFLAGS) == 0):
-        ctx.env.prepend_value('CFLAGS', ["-std=gnu11", "-g", "-Os", "-Wall", "-Wextra", "-Wshadow", "-Wcast-align",
+        ctx.env.prepend_value('CFLAGS', ["-std=gnu11", "-g", "-O0", "-Wall", "-Wextra", "-Wshadow", "-Wcast-align",
```
Comment (yashi) [inline wscript]: this is notiong to do with "removing address", no?

File: examples/python_bindings_example_server.py [hunk 2]
```diff
@@ -126,4 +126,5 @@ def csp_server():
     libcsp.print_routes()
 
     # start CSP server in a thread
-    threading.Thread(target=csp_server).start()
+    #threading.Thread(target=csp_server).start()
+    csp_server()
```
Comment (yashi) [inline examples/python_bindings_example_server.py]: nitpicking: no newline

File: src/bindings/python/pycsp.c [hunk 3]
```diff
@@ -73,10 +73,6 @@ static void pycsp_free_csp_conn(PyObject * obj) {
 }
 
 static void pycsp_free_csp_socket(PyObject * obj) {
-	csp_conn_t * socket = (csp_conn_t *)get_obj_as_socket(obj, true);
-	if (socket) {
-		csp_close(socket);
-	}
```
Comment (yashi) [inline src/bindings/python/pycsp.c]: If you don't close the socket, you are leaking file descritors, no?

File: examples/csp_server_client.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+
```
Comment (yashi) [inline examples/csp_server_client.py:1]: would be nice if you can format according to pep8.

File: wscript [hunk 5]
```diff
@@ -73,7 +73,7 @@ def configure(ctx):
 
     # Setup CFLAGS
     if (len(ctx.stack_path) <= 1) and (len(ctx.env.CFLAGS) == 0):
-        ctx.env.prepend_value('CFLAGS', ["-std=gnu11", "-g", "-O0", "-Wall", "-Wextra", "-Wshadow", "-Wcast-align",
+        ctx.env.prepend_value('CFLAGS', ["-std=gnu11", "-g", "-Os", "-Wall", "-Wextra", "-Wshadow", "-Wcast-align",
```
Comment (yashi) [inline wscript]: This is the change you did in the first commit "remove address from pycsp init".  If you don't need it remove it from the first commit.

File: examples/python_bindings_example_server.py [hunk 6]
```diff
@@ -126,5 +126,4 @@ def csp_server():
     libcsp.print_routes()
 
     # start CSP server in a thread
-    #threading.Thread(target=csp_server).start()
-    csp_server()
\ No newline at end of file
+    threading.Thread(target=csp_server).start()
```
Comment (yashi) [inline examples/python_bindings_example_server.py]: And this is not related to `-O0` removal.

File: examples/csp_server_client.py [hunk 7]
```diff
@@ -54,10 +54,10 @@ def client_task(addr: int) -> None:
         if packet is None:
             raise Exception('Failed to get CSP buffer')
         
-        data = bytes(f'Hello World {chr(count)}', 'ascii') + b'\x00'
+        data = bytearray(f'Hello World {chr(count)}', 'ascii') + b'\x00'
```
Comment (yashi) [inline examples/csp_server_client.py:61]: Why do you need `bytearray` instead of `bytes`?

File: src/bindings/python/pycsp.c [hunk 8]
```diff
@@ -129,10 +129,10 @@ static PyObject * pycsp_socket(PyObject * self, PyObject * args) {
 		return NULL;  // TypeError is thrown
 	}
 
-	csp_socket_t * socket = malloc(sizeof(*socket));
+	csp_socket_t * socket = PyMem_RawMalloc(sizeof(*socket));
```
Comment (yashi) [inline src/bindings/python/pycsp.c:133]: Is this the one you mentioned in the first comment?

> pycsp_socket caused a segmentation fault due to a pointer to a stack allocated socket being used. This was changed to instead use pythons memory allocator for the allocation of the socket.

File: CMakeLists.txt [hunk 9]
```diff
@@ -70,12 +70,14 @@ target_compile_options(csp PRIVATE ${CSP_C_ARGS})
 add_subdirectory(src)
 add_subdirectory(examples)
 
-if(${enable-python3-bindings})
-  find_package(Python COMPONENTS Development.Module)
+if(${CSP_ENABLE_PYTHON3_BINDINGS})
+  find_package(Python COMPONENTS Development)
   if(Python_Development.Module_FOUND)
-    Python_add_library(libcsp_py3 src/bindings/python/pycsp.c WITH_SOABI)
+    Python_add_library(libcsp_py3 STATIC src/bindings/python/pycsp.c)
```
Comment (yashi) [inline CMakeLists.txt]: These lines should be something like:
```cmake
if(${CSP_ENABLE_PYTHON3_BINDINGS})
  find_package(Python3 COMPONENTS Development.Module)
  if(Python3_Development.Module_FOUND)
    Python3_add_library(libcsp_py3 MODULE src/bindings/python/pycsp.c)
```

Because 
1. We don't support python2 any more
2. We are targetting module and module only

**Comment by yashi** [comment]
Overall, thank you for your time on fixing the python binding.  We really appreciate it.  We haven't have good commit messages in libcsp but it'd be really helpful if you can put "why"s in each commit message.
Let me know if you need any help.  Thanks.

File: src/bindings/python/pycsp.c (hunk 3 above)
Comment (sigmundklaa) [reply src/bindings/python/pycsp.c]: Good catch, the port will remain open and associated with the freed socket. `csp_close` is not correct though as that works on `csp_conn_t`, not `csp_sock_t`, so I will make a new function to perform cleanup on a socket.

File: src/bindings/python/pycsp.c (hunk 8 above)
Comment (sigmundklaa) [reply src/bindings/python/pycsp.c:133]: No, it was previously stack-allocated and I changed it to malloc. After looking at the [Python docs](https://docs.python.org/3/c-api/memory.html) I later decided to change it to the Python allocator instead as I figured that was more fitting.

File: examples/csp_server_client.py (hunk 4 above)
Comment (sigmundklaa) [reply examples/csp_server_client.py:1]: Yes, will do

File: examples/csp_server_client.py (hunk 7 above)
Comment (sigmundklaa) [reply examples/csp_server_client.py:61]: `packet_set_data` does not accept read-only bytes object:
Screenshot 2023-07-10 at 09 16 59

File: examples/python_bindings_example_server.py (hunk 2 above)
Comment (sigmundklaa) [reply examples/python_bindings_example_server.py]: Should be fixed with pep8

File: wscript (hunk 5 above)
Comment (sigmundklaa) [reply wscript]: Yes this was accidentally commited, my bad. What do you mean by "remove it from the first commit" though? Is there another way of doing it other than reverting and making a new commit?

File: CMakeLists.txt (hunk 9 above)
Comment (sigmundklaa) [reply CMakeLists.txt]: Updated

File: wscript (hunk 5 above)
Comment (yashi) [reply wscript]: It depends on your git client (aka git porcelain).  A porcelain like [Magit](https://magit.vc/) let you move a hunk from a commit and merge the hunk to the other, which effectively fix this kind of mistake easily.

File: examples/csp_server_client.py (hunk 7 above)
Comment (yashi) [reply examples/csp_server_client.py:61]: Yeah, that's because we have `"Ow*"`.  But to me, It seems the function just does `memcpy()` from the passed argument to `packet->data`.  I don't know why it can't be read-only.

**Comment by sigmundklaa** [comment]
I will look at resolving the rest of the comments next week hopefully, will also try to do better commit names. Agree the commit messages are quite bad so im tempted to make a brand new pull request with better commits, what do you think @yashi ?

**Comment by yashi** [comment]
@sigmundklaa Sure, that's fine by me.  TIA
